flink-runtime/src/main/java/org/apache/flink/runtime/io/network/buffer/BufferCompressor.java

```diff
@@ -0,0 +1,136 @@
+/*
+ * Licensed to the Apache Software Foundation (ASF) under one
+ * or more contributor license agreements.  See the NOTICE file
+ * distributed with this work for additional information
+ * regarding copyright ownership.  The ASF licenses this file
+ * to you under the Apache License, Version 2.0 (the
+ * "License"); you may not use this file except in compliance
+ * with the License.  You may obtain a copy of the License at
+ *
+ *     http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+
+package org.apache.flink.runtime.io.network.buffer;
+
+import org.apache.flink.core.memory.MemorySegment;
+import org.apache.flink.core.memory.MemorySegmentFactory;
+import org.apache.flink.runtime.io.compression.BlockCompressionFactory;
+import org.apache.flink.runtime.io.compression.BlockCompressor;
+
+import java.nio.ByteBuffer;
+
+import static org.apache.flink.util.Preconditions.checkArgument;
+
+/**
+ * Compressor for {@link Buffer}.
+ */
+public class BufferCompressor {
+
+	/** The backing block compressor for data compression. */
+	private final BlockCompressor blockCompressor;
+
+	/** The intermediate heap buffer for the compressed data. */
+	private final byte[] heapBuffer;
+
+	public BufferCompressor(int bufferSize, String factoryName) {
+		checkArgument(bufferSize > 0);
+		// the size of this intermediate heap buffer will be gotten from the
+		// plugin configuration in the future, and currently, double size of
+		// the input buffer is enough for lz4-java compression library.
+		this.heapBuffer = new byte[2 * bufferSize];
+		this.blockCompressor = BlockCompressionFactory.createBlockCompressionFactory(factoryName).getCompressor();
+	}
+
+	/**
+	 * Compresses the given {@link Buffer} using {@link BlockCompressor}. The compressed data will be stored in the
+	 * internal heap buffer of this {@link BufferCompressor} and returned to the caller. The caller must guarantee
+	 * that the returned {@link Buffer} is freed when calling the method next time.
+	 *
+	 * <p>Notes that the compression will always start from offset 0 to the size of the input {@link Buffer}.
+	 */
+	public Buffer compressToInternalBuffer(Buffer buffer) {
+		int compressedLen;
+		if ((compressedLen = compress(buffer)) == 0) {
+			return buffer;
+		}
+
+		try {
+			if (compressedLen >= buffer.getSize()) {
+				return buffer;
+			}
+
+			// warp the internal heap buffer as Buffer
+			MemorySegment memorySegment = MemorySegmentFactory.wrap(heapBuffer);
+			NetworkBuffer compressedBuffer = new NetworkBuffer(memorySegment, FreeingBufferRecycler.INSTANCE);
```

Comment:
We can make the size and compression properties inside the constructor of `NetworkBuffer` to avoid setting them afterwards.